modules/core/src/main/java/org/apache/ignite/internal/util/CompletableFutures.java

```diff
@@ -106,4 +107,26 @@ public static <T> CompletableFuture<List<T>> allOf(CompletableFuture<T>... cfs)
                     return result;
                 });
     }
+
+    /**
+     * Executes the provided function and calls a callback if it fails.
+     *
+     * @param func Function to run.
+     * @param onFailureRun Callback to be invoked after fail.
+     * @param <T> Future result type.
+     * @return A future object representing the result of function call.
+     */
+    public static <T> CompletableFuture<T> doWithCallbackOnFailure(Supplier<CompletableFuture<T>> func, Runnable onFailureRun) {
```

Comment:
I would prefer to remove the function and use `exceptionally` explicitly.

Reply:
Done